test/verify/check-testlib

```diff
@@ -105,6 +106,84 @@ class TestRunTest(MachineCase):
+            return
+
+        env = os.environ.copy()
+        try:
+            env["TEST_FAILURES"] = "1"
+            del env["TEST_JOBS"]
+        except KeyError:
+            pass
+
+        # Check that we retry 3 times failing tests
+        process = subprocess.run([run_tests, "--test-dir", VERIFY_DIR, "TestExample.testFail", "TestExample.testSkip"],
+                                 env=env, capture_output=True)
+        stdout = process.stdout
+        self.assertRegex(stdout, b"\nnot ok 1 .*test\/verify\/check-example TestExample.testFail # RETRY 1 \(be robust against unstable tests\)\n")
+        self.assertRegex(stdout, b"\nnot ok 1 .*test\/verify\/check-example TestExample.testFail # RETRY 2 \(be robust against unstable tests\)\n")
+        self.assertRegex(stdout, b"\nnot ok 1 .*test\/verify\/check-example TestExample.testFail\n")
+        self.assertNotRegex(stdout, b"RETRY 3")
+        self.assertRegex(stdout, b"\nok 2 .*test\/verify\/check-example TestExample.testSkip # SKIP testSkip \(__main__\.TestExample\)\n")
+        self.assertRegex(stdout, b"# 1 TESTS FAILED \[\d*s on .*, 0 serial tests: \]")
+
+        # Check retry logic for changed tests
+        try:  # Wrap in one big try, so when it fails, we still can use finally to cleanup after ourselves
+            with open(os.path.join(VERIFY_DIR, "check-examples"), 'r') as f:
+                original_source = f.read()
+
+            with open(os.path.join(VERIFY_DIR, "check-examples"), 'w') as f:
+                f.write(original_source.replace("NoTest", "test"))
+
+            process = subprocess.run([run_tests, "--test-dir", VERIFY_DIR, "TestRetryExample.testFail", "TestRetryExample.testBasic",
+                                      "TestRetryExample.testNoRetry"], env=env, capture_output=True)
+            stdout = process.stdout
+
+            # Changed test need to succeed 3 times
+            self.assertRegex(stdout, b"\nok 1 .*test\/verify\/check-examples TestRetryExample.testBasic # RETRY 1 \(test affected tests 3 times\)\n")
+            self.assertRegex(stdout, b"\nok 1 .*test\/verify\/check-examples TestRetryExample.testBasic # RETRY 2 \(test affected tests 3 times\)\n")
+            self.assertRegex(stdout, b"\nok 1 .*test\/verify\/check-examples TestRetryExample.testBasic\n")
+            self.assertNotRegex(stdout, b"RETRY 3")
+
+            # Changed test is never retried
+            self.assertRegex(stdout, b"\nnot ok 2 .*test\/verify\/check-examples TestRetryExample.testFail\n")
+            self.assertNotRegex(stdout, b"testFail # RETRY")
+
+            # Using @no_retry_when_changed prevents this retry logic
+            self.assertRegex(stdout, b"\nok 3 .*test\/verify\/check-examples TestRetryExample.testNoRetry\n")
+            self.assertNotRegex(stdout, b"testNoRetry # RETRY")
+        finally:
+            with open(os.path.join(VERIFY_DIR, "check-examples"), 'w') as f:
+                f.write(original_source)
```

Comment:
I recommend to do this with `self.addCleanup()` instead, that's a little more robust and avoids these huge retry blocks.